Add support for signature extensions

adityasaky opened the pull request:
First draft of signature extensions including a stab at defining the sigstore extension in the repo. I think we'd perhaps be better off not listing extensions here.

#59

TomHennen commented on envelope.proto:
```diff
@@ -32,4 +34,19 @@ message Signature {
+message Extension {
+  // Unambiguously identifies how to interpret extension fields.
+  // REQUIRED.
+  string type_ = 1;
```
Does json serialization strip the _?  If not I think you can use the `json_name` extension? https://protobuf.dev/programming-guides/proto3/#json

haydentherapper replied:
I'm not used to seeing fields ending in `_`, can we remove?

woodruffw replied:
Agreed. If the issue here is that `type` might be a keyword in some languages, perhaps we could make it `kind`?

TomHennen commented on extensions/sigstore.md:
```diff
@@ -0,0 +1,51 @@
+
+## Definition
+
+The Sigstore extension's type must be set to `https://sigstore.dev`.
```
nit: Should this be versioned? (IDK)

adityasaky replied:
Yeah, I think it probably should be. I wasn't sure how much to specify here, I wanted to hear from others.

adityasaky replied:
I think using the media type directly (with version) might work best rather than using a URI here and a media type elsewhere.

TomHennen commented on extensions/sigstore.md:
```diff
@@ -0,0 +1,51 @@
+
+FIXME: we need to identify restrictions on the extension's type. Also, what is
+the policy on registering well known extensions? If DSSE points to an external
+definition, we have to ensure the link remains live and correct. On the other
```
Do we have to?  Could we say "you're responsible for ensuring the link is live and correct"?

Maybe when folks submit the links to their specification they should be required to include the repo/bug tracker they use to manage the spec so they can be contacted in case things break?

It does seem similar to what in-toto attestations does (storing them in the repo).  That is a bit of a burden to be sure.

mnm678 replied:
My vote would be just storing links where possible. The full extension definition might change/have new versions, and it's nice to have that in just one place. The links might get out of date, but probably less often than the specification changes.

adityasaky replied:
This makes sense to me. Storing links is fine, for the sigstore example, can we add some mention of this to use the VerificationMaterials definition, @haydentherapper?

haydentherapper replied:
I agree with links. Specification will hopefully come eventually, we can point to that later on.

One question is whether we should point to `main` or a specific commit. I think both come with downsides - main might contain breaking changes not yet reflected in a released version, but a specific commit might ignore the latest updates.

adityasaky replied:
If we have the version recorded here with a link to the sigstore doc, then `main` won't work. The version recorded here must map to the specific definition for that version on the sigstore side so I suspect we're going to have to use the commit (or tag).

adityasaky replied:
Since we want to do links only, I've renamed this folder and created an extensions.md file with a table to list extensions with links. Is that preferable?

TomHennen commented on extensions/sigstore.md:
Should we also request a pointer to code/libraries that can validate the extensions/sigs?

haydentherapper replied:
+1

mnm678 reviewed (approved):
Pending the couple of "FIXME"s this looks good to me!

znewman01 commented:
CC Sigstore clients who might be consumers of this: @woodruffw @haydentherapper @codysoyland @bdehamer (and me, I'll take a look at some point soon)

haydentherapper commented on extensions/sigstore.md:
```diff
@@ -0,0 +1,51 @@
+# Signature Extension: Sigstore
```
Should we add a CODEOWNERS file so that a set of sigstore maintainers are notified when this file is updated?

adityasaky replied:
Sounds reasonable but if we're moving to a common list for multiple extensions, that could get spammy?

haydentherapper commented:
I’m also generally good with this. I do expect some changes once this is actually used though, so I’d prefer we mark it as alpha or wip so it’s easy to make breaking changes.

TomHennen commented:
Is this PR just waiting for me at this point? :)

adityasaky commented:
@TomHennen I'm going to remove the sigstore.md file, I retained it while we checked if just a table was enough. Once I do that and squash my commits, I think we should be good to go with a review from you?

adityasaky commented:
@TomHennen done!

TomHennen reviewed (approved):
Thank you!

haydentherapper commented:
I think we're just waiting on https://github.com/sigstore/protobuf-specs/pull/145?

adityasaky commented:
Yes, sorry, I'm going to be addressing that thread this afternoon / evening and then updating this PR.

woodruffw commented:
Sorry for the belated comment here, but to make sure I understand: the intended outcome here is to allow a DSSE envelope to contain a Sigstore bundle (or whatever else), correct?

In other words: `ext` will contain the bundle's fields, modulo the signature that would otherwise be in the bundle?

adityasaky commented:
Essentially yes, the idea is to associate each _signature_ in a DSSE envelope with its sigstore specific fields (tlog entries etc.).

woodruffw commented:
Okay, thanks! 

This is probably the wrong place to figure this out, but I'm still a little murky on the benefits of having the DSSE as the top-level envelope rather than the Sigstore bundle (or vice versa, except for compatibility reasons in extant Sigstore clients). But that's not a reason for it not to exist, just a lack of clarity on my part 🙂

adityasaky commented:
https://github.com/sigstore/sig-clients/issues/9 might shed some light on that. The tl;dr for the sigstore bundle vs dsse extension:
* a bundle may or may not be a de-facto envelope, depending on whether it embeds a DSSE envelope or a raw signature
* a bundle associates verification information per signature, but DSSE envelopes can contain more than one signature, meaning we'd have to either store the envelope once and somehow infer verification materials for each signature later, or store multiple bundles each with the full envelope but different signatures (and their corresponding verification data)

adityasaky commented:
https://github.com/sigstore/protobuf-specs/pull/145 has been merged, unblocking this. @mnm678 @TomHennen @trishankatdatadog @MarkLodato @JustinCappos

JustinCappos commented on envelope.md:
````diff
@@ -52,14 +52,52 @@ envelopes with individual signatures.
+```
+
+`extension.kind` is a string and the `EXTENSION_KIND` value must unambiguously
+identify the type of the extension. This in turn identifies the fields in the
````
What is the process for obtaining such a string?  

Do we track the set of assigned ones somewhere?

Where are the well-known ones defined?  Can we point to this location in the repo?

adityasaky replied:
I've added some text to extensions.md. I've also linked to that page here.

JustinCappos commented on envelope.md:
```diff
@@ -52,14 +52,52 @@ envelopes with individual signatures.
+listed with their `ext` definition alongside the DSSE specification.
+
+Additionally, extensions must not contain any information such that the
+signature verification fails in its presence and passes in its absence.
```
Is the converse possible?

adityasaky replied:
Do you mean verification passes with the ext but not without? Sure, if a client requires this information, say in the sigstore case. Does this need to be called out here explicitly?

adityasaky replied:
bump!

JustinCappos commented on envelope.md:
```diff
@@ -52,14 +52,52 @@ envelopes with individual signatures.
+Additionally, extensions must not contain any information such that the
+signature verification fails in its presence and passes in its absence.
+Essentially, if a required extension in some context is missing or if a consumer
+does not recognize the extension, verification must fail closed.
```
This makes sense, but what is the implication here for a distributed setting?  It seems like clients without an extension may fail things which clients with the extension may pass.  

Is this intentional?  What does this mean for things like TUF or in-toto which may have may different parties performing verification of the same information?

adityasaky replied:
Yes, for example, if an in-toto client doesn't understand the sigstore extension, it must fail. This is where the producer <-> consumer agreement comes into play to ensure the signing ecosystem is used by the must-pass verifier clients.

adityasaky replied:
Does this need any edits to the doc to resolve?

JustinCappos commented on envelope.md:
```diff
@@ -52,14 +52,52 @@ envelopes with individual signatures.
-    An unset field MUST be treated the same as set-but-empty.
+*   The following fields are OPTIONAL and MAY be unset: `signature.keyid`,
+    `signature.extension`. An unset field MUST be treated the same as
+    set-but-empty.
```
good catch!

JustinCappos commented on protocol.md:
```diff
@@ -52,6 +53,17 @@ KEYID           | string | No       | No
+    describes the fields for the extension. The details for each extension's
+    type and its fields must be agreed upon out-of-band by the signer and
+    verifier, though some well-known extension types may be listed and defined
+    in the DSSE specification. Verifiers MUST use extension fields as a matter
```
Do we want them listed in the specification or a separate page?  I would vote for something separate so we don't have to update the DSSE spec doc for any additions.  Adding an extension should be very lightweight and easy.

adityasaky replied:
It is a separate page, part of this PR. I've included a link.

JustinCappos commented on protocol.md:
```diff
@@ -77,23 +89,27 @@ Functions:
+    EXTENSION such as from the recommended [JSON envelope](envelope.md). Reject
+    if decoding fails.
 -   Optionally, filter acceptable public keys by KEYID.
+-   Optionally, use EXTENSION fields if available to obtain information that may
```
It isn't optional to use the EXTENSION fields, right?  It's just that one may not understand them because they are optional.  In other words, I believe a "correct" client must verify all extension fields it understands.

adityasaky replied:
I think you're right, at least in in-toto/TUF use cases. How should we frame this?

adityasaky replied:
I took a pass at editing this. Please take a look.

adityasaky replied:
How does this read? I'll make the word "may" all caps, but I think it handles @MarkLodato's comment about an extension that may not actually be relevant.

```
-   If EXTENSION is set, use its fields to obtain information that may be
    required for verification. Reject any significant fields that the verifier
    must establish separately, such as the public key for a signature or a root
    certificate for an X.509 certificate chain.
```

JustinCappos replied:
> @JustinCappos I don't see how this follows. Suppose I am expecting a DSSE using a pre-shared key. Someone adds an invalid "Sigstore" extension to the DSSE but doesn't touch any other field, so the DSSE still passes if you ignore the extension. Why would a client be REQUIRED to reject such a DSSE? I can see that they MAY do that, but it doesn't seem mandatory to me.

Would a client that checks the flawed Sigstore extension reject it, even if other things pass?  It just seems odd to me that you could remove or not understand an extension as a way to have something pass.

MarkLodato replied:
Oh, @trishankatdatadog just pointed out that the text already says this:

> Additionally, extensions MUST NOT contain any information such that the
> signature verification fails in its presence and passes in its absence.

It sounds like you are disagreeing with this statement, @JustinCappos?

JustinCappos reviewed (changes requested):
Looks like a great start.  I added some thoughts / requests for minor clarification.

haydentherapper commented on extensions.md:
```diff
@@ -0,0 +1,15 @@
+
+| Name (with link) | Kind | Notes |
+|------------------|------|-------|
+| [Sigstore] | `application/vnd.dev.sigstore.verificationmaterial;version=<VERSION>` | The X.509 certificate chain must not include the root or trusted intermediate certificates |
```
@jku One thing we were discussing is a concern about recursion, where an extension might contain a DSSE envelope which could contain an extension, and so on. For this specific example for Sigstore, this won't be an issue because we're using [VerificationMaterial](https://github.com/sigstore/protobuf-specs/blob/d96fa4407cb4489233f133ab036c0ed98c2756a4/protos/sigstore_bundle.proto#L59) rather than the bundle. The former won't contain a DSSE, the latter would.

I think this is worth noting though, that extensions should (must?) not contain a DSSE to avoid recursion.

Thoughts @adityasaky?

adityasaky replied:
There's a statement about an extension not containing an envelope in envelope.md. Resolving this thread for now.

lukpueh commented on protocol.md:
```diff
@@ -77,23 +89,27 @@ Functions:
 -   Sign PAE(UTF8(PAYLOAD_TYPE), SERIALIZED_BODY). Call the result SIGNATURE.
 -   Optionally, compute a KEYID.
+-   Optionally, include signature specific information as an EXTENSION.
 -   Encode and transmit SERIALIZED_BODY, PAYLOAD_TYPE, SIGNATURE, and KEYID,
```
```suggestion
-   Encode and transmit SERIALIZED_BODY, PAYLOAD_TYPE, SIGNATURE, KEYID and EXTENSION.
```

It is also not mentioned in  "Multi-signature Verification".

lukpueh commented on protocol.md:
```diff
@@ -23,6 +23,7 @@ Name            | Type   | Required | Authenticated
 SERIALIZED_BODY | bytes  | Yes      | Yes
 PAYLOAD_TYPE    | string | Yes      | Yes
 KEYID           | string | No       | No
+EXTENSION       | object | No       | No
```
What does parameter mean here and how does it relate to the SIGNATURE definition above? Is EXTENSION really a parameter or an implementation detail of SIGNATURE?

adityasaky replied:
I think it's the latter but I chose to include this given KEYID is also here. Perhaps we need to restructure these?

lukpueh replied:
Okay, I'm okay with adding it. Would you mind writing an issue?

adityasaky replied:
https://github.com/secure-systems-lab/dsse/issues/66

lukpueh reviewed (changes requested):
This looks great! I'm still a bit fuzzy about how an extension relates to the signature bytes.  The more recent comments from #59 suggest that it should be allowed to use the extension as replacement for the bytes in sig.

adityasaky commented on envelope.md:
```diff
@@ -52,14 +52,52 @@ envelopes with individual signatures.
+field to store ecosystem specific information pertaining to the signature.
+Extensions do not modify the signing workflow established in the
+[DSSE protocol](protocol.md). However, signers and verifiers may exchange
+information about the extensions used out-of-band.
```
Yes, but this information is necessary to establish what extensions a client needs to understand. For example: https://github.com/secure-systems-lab/dsse/pull/61/files#r1445232491

adityasaky commented on envelope.md:
```diff
@@ -52,14 +52,52 @@ envelopes with individual signatures.
+
+In addition to `keyid` and `sig`, a signature object may include an `extension`
+field to store ecosystem specific information pertaining to the signature.
+Extensions do not modify the signing workflow established in the
```
good point, tweaking this.

trishankatdatadog commented:
> I don't think the delta for this is exactly large (or admittedly too small for the current impl), but I think the security gain is worthwhile. What're everybody's thoughts on this?

Agreed. The question is: where? I believe the functionary (in-toto) / role (TUF) that bestows trust on the public key (and hence, the signature) in the first place is the best place to securely bind the extension type (along with other information we already distribute, such as key type, scheme, and so on) to the signature. WDYT?

adityasaky commented:
I think putting it in the policy (in-toto layout, TUF metadata, etc.) is one half of it. I parsed @SantiagoTorres's comment as a recommendation to sign over the extension kind as well, so a signature that uses an extension cannot be verified without explicitly acknowledging the existence of the extension.

This would require a modification to the signing protocol, probably in the PAE?

trishankatdatadog commented:
@adityasaky Yes, you need both for maximum coverage.

SantiagoTorres commented:
yeah, essentially, adding it to the PAE would do the trick, but I think that'd break backwards compat unless you add an if statement to the PAE computation in the case of empty sigs (or an if statement and call this dssev2). I'm curious to know what does @MarkLodato and @haydentherapper think about this tho!

MarkLodato commented:
I just realized that I never recorded my disagreement with the extension mechanism on GitHub (though I had said it during our meeting). I just added https://github.com/secure-systems-lab/dsse/issues/59#issuecomment-1969341050 now. As I said there, my preference would be to merge the Sigstore bundle into DSSE directly so that other non-Sigstore use cases can take advantage. That's why I hadn't been commenting on this PR, since it seemed like I was the only one who disagreed.

That said, I'd recommend against signing the kind. The extension is should just be extra metadata required for verification, no? I don't think it should fundamentally alter the signature?

SantiagoTorres commented:
Thanks for the refresher on #56!

My position on hashing the kind lands around the fact that some security properties could be "peeled off" by a transport (be it malicious or not). One case would be the fact that a sigstore bundle exists alltogether, thus removing sigstore. This, at best, would make a paranoid verifier know that something is iffy (i.e., because it expects the extension), at worse, it'd remove sigstore's security benefits wholesale.

I'm not sure if we want to make it so that verifiers need to enforce policy on extensions existing, and I'm a little nervous about things not being DSSE anymore in that case :)

jkjell commented:
> I'm not sure if we want to make it so that verifiers need to enforce policy on extensions existing

Isn't that what Sigstore is already doing with `cosign`? For example:
```
$ cosign verify <image URI> --certificate-identity=name@example.com
                            --certificate-oidc-issuer=https://accounts.example.com
```

Perhaps there's a difference to "enforce policy" versus "defining a functionary" that I'm conflating. ☺️

trishankatdatadog commented:
> I'm not sure if we want to make it so that verifiers need to enforce policy on extensions existing, and I'm a little nervous about things not being DSSE anymore in that case :)

My concern is that, with either option, DSSE could stop being DS anymore, because, either way, where do you draw the line for more and more extensions?

At least with Option 2, it's sparkling clear how the verifier should handle the envelope. With Option 1, anything goes (especially with an evergrowing combination of extensions) unless there is sparkling clear specification.

What did I miss?

MarkLodato commented:
To me, that's the worst of both worlds. I don't see the signing of the kind providing any security value since the attacker can still modify the bundle, yet it breaks backwards compatibility and adds complexity.

MarkLodato commented:
> the attacker can't forge a bundle if they can't authenticate the bundle themselves and they must provide a bundle.

What is stopping someone from deleting the `tlog_entires` and `timestamp_verification_data` from within the bundle?

SantiagoTorres commented:
> What is stopping someone from deleting the tlog_entires and timestamp_verification_data from within the bundle?

At that point we'd be arguing whether the extension itself should be structured in form x or y (i.e., are `tlog_entries` or `timestamp_verification_data` optional?). We still know we expect a sigstore bundle, however.

MarkLodato commented:
Unless you can point to a specific attack that this would stop for Sigstore (our only use case for extensions thus far), I am strongly against signing the kind. It just adds complexity and breaks backwards compatibility for seemingly no benefit. "Know we expect a sigstore bundle" doesn't really provide any value. The verifier already knows that they need a sigstore bundle; they're not just going to blindly accept any arbitrary signature from any source.

mrjoelkamp commented:
> @SantiagoTorres  I'm not sure if we want to make it so that verifiers need to enforce policy on extensions existing, and I'm a little nervous about things not being DSSE anymore in that case :)

We currently have some prototypes using the DSSE extension as defined in this PR (option 2 in #59) and the enforcement of what extensions the signature is required to have can be handled in a policy (rego) in the client verification process, similarly to how you would require specific attestations to be attached to the image, the policy can require that the DSSE signature have specific extensions. That way you can enforce a transparency log (TL) entry and/or a timestamp authority (TSA) to be attached to the signature. I believe it makes most sense to handle this in policy vs trying to enforce it in DSSE.

SantiagoTorres commented:
>  the enforcement of what extensions the signature is required to have can be handled in a policy (rego) in the client verification process, similarly to how you would require specific attestations to be attached to the image, the policy can require that the DSSE signature have specific extensions.

I  agree that these two are equivalent.  I'm mostly being paranoid about not everybody implementing this check.

To reiterate, I'm not strongly married to the idea of hashing the key, but I wanted to properly discuss it being a possibility. Where I'm coming from, the hashing of the key forces the DSSE layer to ensure the signature itself fails close if a particular extension isn't present.  Instead, an implementer of verification could foot-gun themselves into passing verification if the extension isn't present.

Don't get me wrong, *both* would be a bug, but the latter would make it so that the signature portion is the one that fails, rather than a semantic layer above.

I don't see a lot of excitement/support for this portion, so it's perhaps best to forget about that portion :sweat_smile:

adityasaky commented:
I do like the idea of including the extension's kind in the signature. I think it builds at least some guardrails as someone verifying a sigstore signature must explicitly acknowledge they're verifying a sigstore signature, even if they don't do all the right things at the policy layer above such as checking the subject, verifying inclusion in rekor etc.

What's the best way to unblock this?

lukpueh commented on protocol.md:
```diff
@@ -52,6 +53,17 @@ KEYID           | string | No       | No
+    verifier, though some well-known extension types may be listed and defined
+    in the DSSE specification. Verifiers MUST use extension fields as a matter
+    of convenience and they MUST ensure that extension fields are not relied
+    upon exclusively for security decisions. For example, an extension field
```
ping

adityasaky replied:
done!

lukpueh commented:
I think this is fine. Though it would be nice, if we could find consensus with the original author of the dsse spec. cc @MarkLodato

MarkLodato commented:
The proposal as it is currently written looks good to me. Once it is marked as ready for review and comments are cleaned up, I'm happy to approve.

I don't think signing "kind" is a good idea. To repeat:

1. It breaks backwards compatibility.
2. It adds complexity.
3. It adds almost no security value, as far as I can tell.

I suggest keeping this update limited to just adding an unauthenticated extension. A separate issue could discuss updates to the protocol to make some extensions required, which IMO would require a lot more design.

adityasaky commented:
Note: I'll squash my commits when this is ready to merge, I'm holding off from too many force pushes at this point.

haydentherapper commented:
Not to derail this, but do we have a project in mind within Sigstore that will use this? Sigstore adoption is quite different than when this draft started, and the Sigstore clients currently have all implemented the protobuf-based bundle format. I don't expect we'll see another format introduced to avoid any fracturing of supported formats across clients.

Generally I think this is useful outside of Sigstore too, but just wanted to bring this up given Sigstore is called out explicitly here.

lukpueh commented:
Btw. have we resolved the question, if an extension makes the value of the `sig` bytes obsolete? See https://github.com/secure-systems-lab/dsse/issues/59#issuecomment-1569028836 and following comments.

adityasaky commented:
> Btw. have we resolved the question, if an extension makes the value of the sig bytes obsolete? See https://github.com/secure-systems-lab/dsse/issues/59#issuecomment-1569028836 and following comments.

For the sigstore case, yes. I think that comment was based on the extension being the sigstore bundle proto message rather than the verification_material proto message. Perhaps in the mention of an extension not containing another DSSE envelope (suggested by @haydentherapper IIRC), we should also clarify it can't contain sig bytes either.

> Not to derail this, but do we have a project in mind within Sigstore that will use this? Sigstore adoption is quite different than when this draft started, and the Sigstore clients currently have all implemented the protobuf-based bundle format. I don't expect we'll see another format introduced to avoid any fracturing of supported formats across clients.

Another that comes to mind is spiffe/spire, I think @jkjell may be able to weigh in there.

Re sigstore: I definitely understand not wanting to fracture formats! I think the bundle format is great for the cases where we have a single sigstore signature on a DSSE envelope, while the extension is great for multiple signature scenarios, where only some of them may be sigstore signatures (I can see this happening with in-toto and TUF scenarios). If we end up not signing over the extension kind, it's possible sigstore clients don't need to be aware of the extension right away, with the conversion into the bundle format happening at the layer invoking the client (eg: an in-toto or TUF implementation). The invoking clients still benefit from having extensions for the multi-sig, mixed signing scenarios as they wouldn't have to duplicate envelopes / payloads and reason about all of them. WDYT?

jkjell commented:
As Aditya mentioned, signing with SPIFFE x509-SVID issued cert was another likely extension format. It has a [bundle format](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#6-representation-in-the-spiffe-bundle) defined as well. That would also probably have a timestamp signature associated with it. 

I think the proposal looks good as is, and we should move forward with it.

Besides multiple signatures and multiple extensions, the sigstore verification process is a good example of a reason to not sign over the extension type. There are many possible verification flows for the single extension. What version is the bundle? Is it using a Rekor entry or not? Are you doing an online verification of the CTL?

From a practical standpoint, I can see other consumers, outside of Sigstore (Witness, SLSA, etc), supporting the Sigstore extension-type but, I doubt Sigstore would have any use for supporting any other extension type.

haydentherapper commented:
> If we end up not signing over the extension kind

I might have forgotten, but was this discussed previously? I assume we wouldn't want to sign over the extension to avoid this being a breaking change and mandating clients understand the extension format.

> The invoking clients still benefit from having extensions for the multi-sig, mixed signing scenarios as they wouldn't have to duplicate envelopes / payloads and reason about all of them. WDYT?

I agree that I think this approach is better for multi-sigs given the current Sigstore bundle, as the bundle doesn't support a mapping between sigs and verification keys/certs. It's a little hard for me to predict what the exact use case currently since we haven't had a need for multi-sig DSSEs, but we should probably document somewhere that if the need does arise, we should revisit using this format.

trishankatdatadog commented:
> I might have forgotten, but was this discussed previously? I assume we wouldn't want to sign over the extension to avoid this being a breaking change and mandating clients understand the extension format.
> 

If signing over kind is a breaking change, then it should be a V2 requirement (if at all) @SantiagoTorres @JustinCappos

adityasaky commented:
> If signing over kind is a breaking change, then it should be a V2 requirement (if at all)

To clarify, this PR does not propose signing over the extension kind, though it was discussed in the comments. As it stands, this PR does not introduce a breaking change, and one of the points against signing over the kind was to avoid breaking backwards compatibility.

I propose we move ahead with this PR (pending reviews from @SantiagoTorres, @JustinCappos, @lukpueh) but hold off on creating a release of the specification. Before we actually release with extensions, we can nail down whether we want to sign over the kind + anything else that needs to be discussed. Thoughts? :smile:

trishankatdatadog commented:
> I propose we move ahead with this PR (pending reviews from @SantiagoTorres, @JustinCappos, @lukpueh) but hold off on creating a release of the specification. Before we actually release with extensions, we can nail down whether we want to sign over the kind + anything else that needs to be discussed. Thoughts?

That might be too confusing to casual observers who need to realise that just because something is there in a commit, doesn't mean it's been released. I argue for pushing for consensus (been too long, sorry on behalf of everyone) and releasing ASAP. Any objections from [the other 8 maintainers](https://github.com/secure-systems-lab/dsse/blob/f5477864582607ee55c21a7b1b9341163719ec0c/MAINTAINERS.md)?

MarkLodato commented:
> That might be too confusing to casual observers who need to realise that just because something is there in a commit, doesn't mean it's been released. I argue for pushing for consensus (been too long, sorry on behalf of everyone) and releasing ASAP. Any objections from [the other 8 maintainers](https://github.com/secure-systems-lab/dsse/blob/f5477864582607ee55c21a7b1b9341163719ec0c/MAINTAINERS.md)?

I agree. We don't have a formal release process, so we're currently living at head. Submitted == released. Perhaps an alternative is marking the `extension` field as "experimental", with a warning that the design is subject to change? I don't foresee any changes, but it probably is a good idea to get some real-world feedback before committing.

MarkLodato commented on envelope.md:
```diff
@@ -1,8 +1,8 @@
+May 10, 2024
 
-Version 1.0.0
+Version 1.1.0-draft
```
But how are users to find the non-draft version? We don't have any convention for using (say) branches, and there's no documentation explaining this.

I think we need to either:
- (status quo) Say that anything living on `main` is "published", and commit to supporting it long-term.
- Come up with a way to publish draft versions, such as through git branches or by publishing on a website (most likely through GitHub Pages or Netlify).

adityasaky replied:
I think there's been drift in the status quo over time, because there was an explicit tag (https://github.com/secure-systems-lab/dsse/releases/tag/v1.0.0) for v1.0.0 with no patch releases for minor fixes since then. Do we commit to publishing on `main` henceforth and drop explicit tags? I do like the idea of having the tags, and perhaps we ought to clarify in the README to refer to them for releases?

MarkLodato replied:
That would be fine too. It just needs to be documented and clearly communicated to readers. We would also want a branch for "latest" so that users don't always have to look up by tag. I think the cleanest solution is for `main` to always refer to the most recent version and `devel` (or similar) to point to the upcoming draft.

adityasaky replied:
That sounds reasonable to me. Here's what I propose:
* I'll open a separate PR against master that adds a note to the README about versioning
* I'll bump the versions of files on master now (before this PR) to v1.0.1 for the clarifications that have been made since v1.0.0 (I think a patch release is likely fine, skimming through the commits since then)

With this merged, we can tag v1.0.1 against master, which aligns with what you propose. I can then rebase this PR, drop the -draft from the docs, and change the base branch to devel.

trishankatdatadog commented on protocol.md:
```diff
@@ -52,6 +53,20 @@ KEYID           | string | No       | No
+    though some well-known extensions may be [listed](extensions.md) and defined
+    in the DSSE specification. Note that as the extension is unauthenticated,
+    it MUST NOT allow the verifier to independently verify the signature. For
+    example, the extension MUST NOT be trusted to provide the public key to
```
```suggestion
    example, the extension MUST NOT be trusted to directly provide the leaf public key to
```

trishankatdatadog commented on protocol.md:
```diff
@@ -52,6 +53,20 @@ KEYID           | string | No       | No
+    example, the extension MUST NOT be trusted to provide the public key to
+    verify a signature. Similarly, an extension field that contains an X.509
+    certificate chain MUST NOT be trusted to provide the root certificate, but
+    only intermediate certificates.
```
```suggestion
    it may provide only intermediate certificates.
```

adityasaky commented:
As discussed in https://github.com/secure-systems-lab/dsse/pull/61#discussion_r1596835160, #67 has been merged with a versioning scheme and v1.0.1 has been tagged. I've updated this PR to specify the version as just v1.1.0 in each .md file, and I've updated the base branch to be `devel`, which points to the current tip of the `master` branch.

trishankatdatadog reviewed (approved):
Conditional on this key text:

> Additionally, extensions MUST NOT contain any information such that the
signature verification fails in its presence and passes in its absence.

I am approving this PR.

adityasaky commented:
Are we ready to merge this into devel? I see the outstanding review request is for @SantiagoTorres. :smile:

adityasaky commented:
Thank you everyone!
